### PR TITLE
Add symmetrize_from_upper_tri

### PR DESCRIPTION
**Pull request by gregorp90**
## Summary

Added symmetrize_from_upper_tri.cpp function, which creates a symmetric matrix by using elements in the upper triangular part of the input matrix. Implementation is based on the existing symmetrize_from_lower_tri.cpp.

Changed the second test in symmetrize_from_lower_tri_test.cpp. Before, the input was a symmetric matrix. Now, the input is a non-symmetric square matrix.

## Tests

Added 2 unit tests for symmetrize_from_upper_tri.cpp. First one receives invalid input (non-square matrix) and throws an invalid argument error. The other one receives a non-symmetric square matrix and check if the output is symmetric. The tests are based on symmetrize_from_lower_tri_test.cpp.

## Side Effects

None.

## Release notes

Added symmetrize_from_upper_tri() function.

## Checklist

- [x] Math issue #482

- [x] Copyright holder: Gregor Pirš

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
@spinkney would you be up for reviewing this one? No rush, whenever you have time.

File: test/unit/math/prim/fun/symmetrize_from_lower_tri_test.cpp [hunk 1]
```diff
@@ -22,6 +22,6 @@ TEST(Symmetrize, Value) {
   Matrix<double, Dynamic, Dynamic> m(4, 4);
   for (int i = 0; i < 4; ++i)
     for (int j = 0; j < 4; ++j)
-      m(i, j) = (i + 1) * (j + 1);
+      m(i, j) = (i + 2) * (j + 1);
```
Comment (bbbales2) [inline test/unit/math/prim/fun/symmetrize_from_lower_tri_test.cpp:25]: Ooooh, we should be checking that this function gives the right result.

So build the result:
```
Eigen::MatrixXd m_ref(4, 4);
m_ref << a, b, c, d,
    b, e, f, ...;
EXPECT_MATRIX_EQ(m_sym_ref, symmetrize_from_lower_tri(m));
```

File: test/unit/math/prim/fun/symmetrize_from_upper_tri_test.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using stan::math::is_symmetric;
+  using stan::math::symmetrize_from_upper_tri;
+
+  Matrix<double, Dynamic, Dynamic> m(4, 4);
+  for (int i = 0; i < 4; ++i)
+    for (int j = 0; j < 4; ++j)
+      m(i, j) = (i + 2) * (j + 1);
+  EXPECT_TRUE(is_symmetric(symmetrize_from_upper_tri(m)));
```
Comment (bbbales2) [inline test/unit/math/prim/fun/symmetrize_from_upper_tri_test.cpp:26]: Check that this is actually returning the right values too.

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Even though this function is only in prim, still need to add mixed mode tests (to make sure all the higher order autodiff is working -- it is possible it breaks even if this is fully templated).

Check here: https://github.com/stan-dev/math/blob/develop/test/unit/math/mix/fun/rows_dot_product_test.cpp

But don't do any of the `_matvar` tests cause those are for the varmat types (yeah the naming is a bit funny there lol).

File: test/unit/math/prim/fun/symmetrize_from_lower_tri_test.cpp (hunk 1 above)
Comment (gregorp90) [reply test/unit/math/prim/fun/symmetrize_from_lower_tri_test.cpp:25]: Thanks Ben, I'll fix these issues.

**Comment by rok-cesnovar** [comment]
Since this is the first time you are working with the Stan jenkins test, let me give a quick quide.

image

Click "Details" on the failed test.

It seems cpplint has failed

image

Looking at the complete log here https://jenkins.mc-stan.org/blue/rest/organizations/jenkins/pipelines/Math%20Pipeline/branches/PR-2354/runs/7/nodes/53/steps/56/log/?start=0
we find.
```
Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
```
So you need to add a newline.

You can run cpplint tests localy with

```
mingw32-make cpplint
```

**Comment by gregorp90** [comment]
> Since this is the first time you are working with the Stan jenkins test, let me give a quick quide.
> 
> image
> 
> Click "Details" on the failed test.
> 
> It seems cpplint has failed
> 
> image
> 
> Looking at the complete log here https://jenkins.mc-stan.org/blue/rest/organizations/jenkins/pipelines/Math%20Pipeline/branches/PR-2354/runs/7/nodes/53/steps/56/log/?start=0
> we find.
> 
> ```
> Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
> t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
> Done processing test/unitest/unit/math/mix/fun/symmetrize_from_lower_tri_test.cpp(36): error cpplint: [whitespace/ending_newline] Could not find a newline character at the end of the file. [5]
> t/math/mix/fun/symmetrize_from_upper_tri_test.cpp
> ```
> 
> So you need to add a newline.
> 
> You can run cpplint tests localy with
> 
> ```
> mingw32-make cpplint
> ```

**Comment by gregorp90** [comment]
Pull request is ready for review. I made the following changes:

1. Added tests to symmetrize_from_lower_tri and symmetrize_from_upper_tri that check if the function is returning the correct values.
2. Added automatic differentiation tests for symmetrize_from_lower_tri and symmetrize_from_upper_tri.

**Review by bbbales2 — APPROVED** [review]

**Comment by bbbales2** [comment]
Looks good to me. I simplified the mix tests a little. Merge if tests pass and you're happy with my changes to the tests.

**Review by bbbales2 — APPROVED** [review]
Good

**Comment by rok-cesnovar** [comment]
@bbbales2 Gregor does not have permissions here so you can go ahead and merge yourself unless I beat you to it for that sweet sweet Github points. 

@gregorp90 Thanks! And congrats on your first merged pull request to Stan Math! :tada: :beer:

**Comment by gregorp90** [comment]
@bbbales2, I think the word you were looking for is "somewhat", not "little". Thanks again for reviewing and the comments/simplifications! Really helped in my understanding of the project.

Thanks @rok-cesnovar!

**Pull request merged**